These are all the packages you’ll need for Python statistics calculations

In [5]:
import math
import statistics
import numpy as np
import scipy.stats
import pandas as pd

from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

Let’s create some data to work with. You’ll start with Python lists that contain some arbitrary numeric data:

In [3]:
x = [8.0, 1, 2.5, 4, 28.0]

x_with_nan = [8.0, 1, 2.5, math.nan, 4, 28.0]

#x_with_nan contains a nan value. It’s important 
#to understand the behavior of the Python statistics routines when they come across a not-a-number value (nan). 

 How do you get a nan value?

In Python, you can use any of the following:

- float('nan')
- math.nan
- np.nan


### Now, create np.ndarray and pd.Series objects that correspond to x and x_with_nan:



In [4]:
y, y_with_nan = np.array(x), np.array(x_with_nan)


z, z_with_nan = pd.Series(x), pd.Series(x_with_nan)

#### Measures of Central Tendency

The measures of central tendency show the central or middle values of datasets. There are several definitions of what’s considered to be the center of a dataset. In this tutorial, you’ll learn how to identify and calculate these measures of central tendency:

- Mean
- Weighted mean
- Geometric mean
- Harmonic mean
- Median
- Mode

**Mean**

In [13]:
#You can calculate the mean with pure Python using sum() and len(), without importing libraries:
mean_ = sum(x) / len(x)
mean_


#Although this is clean and elegant, you can also apply built-in Python statistics functions:

mean_=statistics.mean(x)
mean_

#If you use NumPy, then you can get the mean with np.mean():

mean_=np.mean(y)
mean_

mean_ = y.mean()
mean_

#The function mean() and method .mean() from NumPy return the same result as statistics.mean(). 
#This is also the case when there are nan values among your data:

np.mean(y_with_nan)

y_with_nan.mean()


#If you prefer to ignore nan values, then you can use np.nanmean():
np.nanmean(y_with_nan)



#pd.Series objects also have the method .mean():As you can see, it’s used similarly as in the case of NumPy. 
#However, .mean() from Pandas ignores nan values by default:
z_with_nan.mean()



8.7

8.7

8.7

8.7

nan

nan

8.7

8.7

#### Weighted Mean

The weighted mean, also called the weighted arithmetic mean or weighted average, is a generalization of the arithmetic mean that enables you to define the relative contribution of each data point to the result.

The weighted mean is very handy when you need the mean of a dataset containing items that occur with given relative frequencies. For example, say that you have a set in which 20% of all items are equal to 2, 50% of the items are equal to 4, and the remaining 30% of the items are equal to 8. You can calculate the mean of such a set like this:


0.2 * 2 + 0.5 * 4 + 0.3 * 8
4.8


You can implement the weighted mean in pure Python by combining sum() with either range() or zip():



In [17]:
x = [8.0, 1, 2.5, 4, 28.0]

w = [0.1, 0.2, 0.3, 0.25, 0.15]


wmean = sum(w[i] * x[i] for i in range(len(x))) / sum(w)

wmean

#or

wmean = sum(x_ * w_ for (x_, w_) in zip(x, w)) / sum(w)

wmean


#However, if you have large datasets, then NumPy is likely to provide a better solution. 
#You can use np.average() to get the weighted mean of NumPy arrays or Pandas Series:


y, z, w = np.array(x), pd.Series(x), np.array(w)

wmean = np.average(y, weights=w)

wmean

#or

wmean = np.average(z, weights=w)

wmean


#Another solution is to use the element-wise product w * y with np.sum() or .sum():

(w * y).sum() / w.sum()

6.95

6.95

6.95

6.95

6.95

### Harmonic Mean

The harmonic mean is the reciprocal of the mean of the reciprocals of all items in the dataset: 𝑛 / Σᵢ(1/𝑥ᵢ), where 𝑖 = 1, 2, …, 𝑛 and 𝑛 is the number of items in the dataset 𝑥. One variant of the pure Python implementation of the harmonic mean is this:

In [20]:
hmean = len(x) / sum(1 / item for item in x)

hmean

#or

hmean = statistics.harmonic_mean(x)

hmean

#A third way to calculate the harmonic mean is to use scipy.stats.hmean():

scipy.stats.hmean(y)

scipy.stats.hmean(z)


2.7613412228796843

2.7613412228796843

2.7613412228796843

2.7613412228796843

#### Geometric Mean

The geometric mean is the 𝑛-th root of the product of all 𝑛 elements 𝑥ᵢ in a dataset 𝑥: ⁿ√(Πᵢ𝑥ᵢ), where 𝑖 = 1, 2, …, 𝑛. The following figure illustrates the arithmetic, harmonic, and geometric means of a dataset:

As you can see, the value of the geometric mean, in this case, differs significantly from the values of the arithmetic (8.7) and harmonic (2.76) means for the same dataset x.

Python 3.8 introduced statistics.geometric_mean(), which converts all values to floating-point numbers and returns their geometric mean:

In [23]:
gmean = 1

for item in x:
     gmean *= item

gmean **= 1 / len(x)

gmean



#or

gmean = statistics.geometric_mean(x)

gmean


## You can also get the geometric mean with scipy.stats.gmean():

scipy.stats.gmean(y)

scipy.stats.gmean(z)


4.677885674856041

4.67788567485604

4.67788567485604

4.67788567485604

### Median

The sample median is the middle element of a sorted dataset. The dataset can be sorted in increasing or decreasing order. If the number of elements 𝑛 of the dataset is odd, then the median is the value at the middle position: 0.5(𝑛 + 1). If 𝑛 is even, then the median is the arithmetic mean of the two values in the middle, that is, the items at the positions 0.5𝑛 and 0.5𝑛 + 1.

For example, if you have the data points 2, 4, 1, 8, and 9, then the median value is 4, which is in the middle of the sorted dataset (1, 2, 4, 8, 9). If the data points are 2, 4, 1, and 8, then the median is 3, which is the average of the two middle elements of the sorted sequence (2 and 4). 

In [26]:
n = len(x)
if n % 2:
    median_ = sorted(x)[round(0.5*(n-1))]
else:
    x_ord, index = sorted(x), round(0.5 * n)
    median_ = 0.5 * (x_ord[index-1] + x_ord[index])

median_



#Two most important steps of this implementation are as follows:

    #Sorting the elements of the dataset
    #Finding the middle element(s) in the sorted dataset

    
#You can get the median with statistics.median():

median_ = statistics.median(x)
median_

median_ = statistics.median(x[:-1])

median_

# The sorted version of x is [1, 2.5, 4, 8.0, 28.0], so the element in the middle is 4. 
#The sorted version of x[:-1], which is x without the last item 28.0, is [1, 2.5, 4, 8.0]. 
# Now, there are two middle elements, 2.5 and 4. Their average is 3.25.


4

4

3.25

median_low() and median_high() are two more functions related to the median in the Python statistics library. They always return an element from the dataset:

If the number of elements is odd, then there’s a single middle value, so these functions behave just like median().

If the number of elements is even, then there are two middle values. In this case, median_low() returns the lower and median_high() the higher middle value.

You can use these functions just as you’d use median():

In [27]:
statistics.median_low(x[:-1])

statistics.median_high(x[:-1])

# Again, the sorted version of x[:-1] is [1, 2.5, 4, 8.0]. The two elements in the middle are 2.5 (low) and 4 (high).

2.5

4

In [29]:
#Unlike most other functions from the Python statistics library, median(), median_low(), and median_high()
#don’t return nan when there are nan values among the data points:

statistics.median(x_with_nan)

statistics.median_low(x_with_nan)

statistics.median_high(x_with_nan)


#Beware of this behavior because it might not be what you want!

#You can also get the median with np.median():

median_ = np.median(y)

median_

median_ = np.median(y[:-1])

median_



6.0

4

8.0

4.0

3.25

In [31]:
#However, if there’s a nan value in your dataset, then np.median() issues the RuntimeWarning and returns nan. 
#If this behavior is not what you want, then you can use nanmedian() to ignore all nan values:

np.nanmedian(y_with_nan)

np.nanmedian(y_with_nan[:-1])


#Pandas Series objects have the method .median() that ignores nan values by default:

z.median()

z_with_nan.median()

4.0

3.25

4.0

4.0

### Mode

The sample mode is the value in the dataset that occurs most frequently. If there isn’t a single such value, then the set is multimodal since it has multiple modal values. For example, in the set that contains the points 2, 3, 2, 8, and 12, the number 2 is the mode because it occurs twice, unlike the other items that occur only once.

This is how you can get the mode with pure Python:

In [41]:
u = [2, 3, 2, 8, 12]

mode_ = max((u.count(item), item) for item in set(u))[1]

mode_

#You use u.count() to get the number of occurrences of each item in u

u.count(12)

set(u)
# set(u) returns a Python set with all unique items in u. 
# You can use this trick to optimize working with larger data, especially when you expect to see a lot of duplicates.

2

1

{2, 3, 8, 12}

In [43]:
mode_ = statistics.mode(u)

mode_

mode_ = statistics.multimode(u)

mode_


#As you can see, mode() returned a single value, while multimode() returned the list that contains the result. 
#This isn’t the only difference between the two functions, though. 
#If there’s more than one modal value, then mode() raises StatisticsError, 
#while multimode() returns the list with all modes:

v = [12, 15, 12, 15, 21, 15, 12]

statistics.mode(v)  # Raises StatisticsError

statistics.multimode(v)


2

[2]

12

[12, 15]

In [44]:
#nan

statistics.mode([2, math.nan, 2])

statistics.multimode([2, math.nan, 2])

statistics.mode([2, math.nan, 0, math.nan, 5])

statistics.multimode([2, math.nan, 0, math.nan, 5])

2

[2]

nan

[nan]

In [45]:
#You can also get the mode with scipy.stats.mode():

u, v = np.array(u), np.array(v)

mode_ = scipy.stats.mode(u)

mode_

mode_ = scipy.stats.mode(v)

mode_


ModeResult(mode=array([2]), count=array([2]))

ModeResult(mode=array([12]), count=array([3]))

In [46]:
#You can get the mode and its number of occurrences as NumPy arrays with dot notation:

mode_.mode

mode_.count


array([12])

array([3])

In [47]:
#Pandas Series objects have the method .mode() that handles multimodal values well and ignores nan values by default:

u, v, w = pd.Series(u), pd.Series(v), pd.Series([2, 2, math.nan])

u.mode()



v.mode()



w.mode()


0    2
dtype: int32

0    12
1    15
dtype: int32

0    2.0
dtype: float64

### Measures of Variability

The measures of central tendency aren’t sufficient to describe data. You’ll also need the measures of variability that quantify the spread of data points. In this section, you’ll learn how to identify and calculate the following variability measures:

- Variance
- Standard deviation
- Skewness
- Percentiles
- Ranges
- Variance

**The sample variance quantifies the spread of the data. It shows numerically how far the data points are from the mean.** You can express the sample variance of the dataset 𝑥 with 𝑛 elements mathematically as 𝑠² = Σᵢ(𝑥ᵢ − mean(𝑥))² / (𝑛 − 1), where 𝑖 = 1, 2, …, 𝑛 and mean(𝑥) is the sample mean of 𝑥. If you want to understand deeper why you divide the sum with 𝑛 − 1 instead of 𝑛, then you can dive deeper into Bessel’s correction.

The following figure shows you why it’s important to consider the variance when describing datasets:

In [51]:
#Here’s how you can calculate the sample variance with pure Python:

n = len(x)

mean_ = sum(x) / n

var_ = sum((item - mean_)**2 for item in x) / (n - 1)

var_


#This approach is sufficient and calculates the sample variance well. 
#However, the shorter and more elegant solution is to call the existing function statistics.variance():

var_ = statistics.variance(x)

var_


#If you have nan values among your data, then statistics.variance() will return nan:

statistics.variance(x_with_nan)


#You can also calculate the sample variance with NumPy. 
#You should use the function np.var() or the corresponding method .var():

var_ = np.var(y, ddof=1)

var_

var_ = y.var(ddof=1)

var_

# It’s very important to specify the parameter ddof=1. That’s how you set the delta degrees of freedom to 1. 
# This parameter allows the proper calculation of 𝑠², with (𝑛 − 1) in the denominator instead of 𝑛.

123.19999999999999

123.2

nan

123.19999999999999

123.19999999999999

In [53]:
#If you have nan values in the dataset, then np.var() and .var() will return nan:

np.var(y_with_nan, ddof=1)


y_with_nan.var(ddof=1)


#This is consistent with np.mean() and np.average(). If you want to skip nan values, then you should use np.nanvar():
np.nanvar(y_with_nan, ddof=1)

nan

nan

123.19999999999999

In [54]:
#pd.Series objects have the method .var() that skips nan values by default:

z.var(ddof=1)

z_with_nan.var(ddof=1)


123.19999999999999

123.19999999999999

### Standard Deviation

The sample standard deviation is another measure of data spread. It’s connected to the sample variance, as standard deviation, 𝑠, is the positive square root of the sample variance. The standard deviation is often more convenient than the variance because it has the same unit as the data points. Once you get the variance, you can calculate the standard deviation with pure Python:

In [55]:
std_ = var_ ** 0.5

std_

11.099549540409285

In [56]:
#Although this solution works, you can also use statistics.stdev():

std_ = statistics.stdev(x)

std_

11.099549540409287

You can get the standard deviation with NumPy in almost the same way. You can use the function std() and the corresponding method .std() to calculate the standard deviation. If there are nan values in the dataset, then they’ll return nan. To ignore nan values, you should use np.nanstd(). You use std(), .std(), and nanstd() from NumPy as you would use var(), .var(), and nanvar():

In [57]:
np.std(y, ddof=1)

y.std(ddof=1)

np.std(y_with_nan, ddof=1)

y_with_nan.std(ddof=1)

np.nanstd(y_with_nan, ddof=1)

11.099549540409285

11.099549540409285

nan

nan

11.099549540409285

In [58]:
#Don’t forget to set the delta degrees of freedom to 1!

#pd.Series objects also have the method .std() that skips nan by default:

z.std(ddof=1)

z_with_nan.std(ddof=1)

# The parameter ddof defaults to 1, so you can omit it.
# Again, if you want to treat nan values differently, then apply the parameter skipna.


11.099549540409285

11.099549540409285

### Skewness

The sample skewness measures the asymmetry of a data sample.

There are several mathematical definitions of skewness. One common expression to calculate the skewness of the dataset 𝑥 with 𝑛 elements is (𝑛² / ((𝑛 − 1)(𝑛 − 2))) (Σᵢ(𝑥ᵢ − mean(𝑥))³ / (𝑛𝑠³)). A simpler expression is Σᵢ(𝑥ᵢ − mean(𝑥))³ 𝑛 / ((𝑛 − 1)(𝑛 − 2)𝑠³), where 𝑖 = 1, 2, …, 𝑛 and mean(𝑥) is the sample mean of 𝑥. The skewness defined like this is called the adjusted Fisher-Pearson standardized moment coefficient.

In [59]:
#Once you’ve calculated the size of your dataset n, the sample mean mean_, and the standard deviation std_, 
#you can get the sample skewness with pure Python:

x = [8.0, 1, 2.5, 4, 28.0]

n = len(x)

mean_ = sum(x) / n

var_ = sum((item - mean_)**2 for item in x) / (n - 1)

std_ = var_ ** 0.5

skew_ = (sum((item - mean_)**3 for item in x)
         * n / ((n - 1) * (n - 2) * std_**3))
skew_


#The skewness is positive, so x has a right-side tail.

1.9470432273905929

In [60]:
#You can also calculate the sample skewness with scipy.stats.skew():

y, y_with_nan = np.array(x), np.array(x_with_nan)

scipy.stats.skew(y, bias=False)

scipy.stats.skew(y_with_nan, bias=False)

#The obtained result is the same as the pure Python implementation.
#The parameter bias is set to False to enable the corrections for statistical bias. 
#The optional parameter nan_policy can take the values 'propagate', 'raise', or 'omit'.
#It allows you to control how you’ll handle nan values.


1.9470432273905927

nan

In [61]:
#Pandas Series objects have the method .skew() that also returns the skewness of a dataset:

z, z_with_nan = pd.Series(x), pd.Series(x_with_nan)

z.skew()


z_with_nan.skew()

# Like other methods, .skew() ignores nan values by default, 
# because of the default value of the optional parameter skipna.

1.9470432273905924

1.9470432273905924

### Percentiles

The sample 𝑝 percentile is the element in the dataset such that 𝑝% of the elements in the dataset are less than or equal to that value. Also, (100 − 𝑝)% of the elements are greater than or equal to that value. If there are two such elements in the dataset, then the sample 𝑝 percentile is their arithmetic mean. Each dataset has three quartiles, which are the percentiles that divide the dataset into four parts:

The first quartile is the sample 25th percentile. It divides roughly 25% of the smallest items from the rest of the dataset.

The second quartile is the sample 50th percentile or the median. Approximately 25% of the items lie between the first and second quartiles and another 25% between the second and third quartiles.

The third quartile is the sample 75th percentile. It divides roughly 25% of the largest items from the rest of the dataset.
Each part has approximately the same number of items. If you want to divide your data into several intervals, then you can use statistics.quantiles():

In [62]:
x = [-5.0, -1.1, 0.1, 2.0, 8.0, 12.8, 21.0, 25.8, 41.0]

statistics.quantiles(x, n=2)

statistics.quantiles(x, n=4, method='inclusive')

# In this example, 8.0 is the median of x, while 0.1 and 21.0 are the sample 25th and 75th percentiles, respectively.
# The parameter n defines the number of resulting equal-probability percentiles, and method determines how to calculate them.

[8.0]

[0.1, 8.0, 21.0]

In [63]:
#You can also use np.percentile() to determine any sample percentile in your dataset. 
#For example, this is how you can find the 5th and 95th percentiles:

y = np.array(x)

np.percentile(y, 5)

np.percentile(y, 95)


# percentile() takes several arguments. 
# You have to provide the dataset as the first argument and the percentile value as the second. 
# The dataset can be in the form of a NumPy array, list, tuple, or similar data structure. 
# The percentile can be a number between 0 and 100 like in the example above, but it can also be a sequence of numbers:



-3.44

34.919999999999995

In [65]:
np.percentile(y, [25, 50, 75])

np.median(y)

#This code calculates the 25th, 50th, and 75th percentiles all at once. 
#If the percentile value is a sequence, then percentile() returns a NumPy array with the results. 
#The first statement returns the array of quartiles. 
#The second statement returns the median, so you can confirm it’s equal to the 50th percentile, which is 8.0.




#If you want to ignore nan values, then use np.nanpercentile() instead:


y_with_nan = np.insert(y, 2, np.nan)

y_with_nan

np.nanpercentile(y_with_nan, [25, 50, 75])

#That’s how you can avoid nan values.

array([ 0.1,  8. , 21. ])

8.0

array([-5. , -1.1,  nan,  0.1,  2. ,  8. , 12.8, 21. , 25.8, 41. ])

array([ 0.1,  8. , 21. ])

In [66]:
#NumPy also offers you very similar functionality in quantile() and nanquantile(). 
#If you use them, then you’ll need to provide the quantile values as the numbers between 0 and 1 instead of percentiles:

np.quantile(y, 0.05)

np.quantile(y, 0.95)

np.quantile(y, [0.25, 0.5, 0.75])

np.nanquantile(y_with_nan, [0.25, 0.5, 0.75])


# The results are the same as in the previous examples, 
# but here your arguments are between 0 and 1. In other words, you passed 0.05 instead of 5 and 0.95 instead of 95.

-3.44

34.919999999999995

array([ 0.1,  8. , 21. ])

array([ 0.1,  8. , 21. ])

In [67]:
#pd.Series objects have the method .quantile():

z, z_with_nan = pd.Series(y), pd.Series(y_with_nan)

z.quantile(0.05)

z.quantile(0.95)



z.quantile([0.25, 0.5, 0.75])


z_with_nan.quantile([0.25, 0.5, 0.75])

#.quantile() also needs you to provide the quantile value as the argument. This value can be a number between 0 and 1 or a sequence of numbers. In the first case, .quantile() returns a scalar. In the second case, it returns a new Series holding the results.



-3.44

34.919999999999995

0.25     0.1
0.50     8.0
0.75    21.0
dtype: float64

0.25     0.1
0.50     8.0
0.75    21.0
dtype: float64

### Ranges

The range of data is the difference between the maximum and minimum element in the dataset. You can get it with the function np.ptp():

In [68]:
np.ptp(y)

np.ptp(z)

np.ptp(y_with_nan)

np.ptp(z_with_nan)


46.0

46.0

nan

nan

This function returns nan if there are nan values in your NumPy array. If you use a Pandas Series object, then it will return a number.

Alternatively, you can use built-in Python, NumPy, or Pandas functions and methods to calculate the maxima and minima of sequences:

- max() and min() from the Python standard library
- amax() and amin() from NumPy
- nanmax() and nanmin() from NumPy to ignore nan values
- .max() and .min() from NumPy
- .max() and .min() from Pandas to ignore nan values by default

Here are some examples of how you would use these routines:

In [69]:
np.amax(y) - np.amin(y)


np.nanmax(y_with_nan) - np.nanmin(y_with_nan)

y.max() - y.min()

z.max() - z.min()

z_with_nan.max() - z_with_nan.min()

#That’s how you get the range of data.


46.0

46.0

46.0

46.0

46.0

In [70]:
#The interquartile range is the difference between the first and third quartile. 
#Once you calculate the quartiles, you can take their difference:

quartiles = np.quantile(y, [0.25, 0.75])
quartiles[1] - quartiles[0]



quartiles = z.quantile([0.25, 0.75])
quartiles[0.75] - quartiles[0.25]


20.9

20.9

### Summary of Descriptive Statistics

SciPy and Pandas offer useful routines to quickly get descriptive statistics with a single function or method call. You can use scipy.stats.describe() like this:

In [73]:
result = scipy.stats.describe(y, ddof=1, bias=False)

result

DescribeResult(nobs=9, minmax=(-5.0, 41.0), mean=11.622222222222222, variance=228.75194444444446, skewness=0.9249043136685094, kurtosis=0.14770623629658886)

describe() returns an object that holds the following descriptive statistics:

- nobs: the number of observations or elements in your dataset
- minmax: the tuple with the minimum and maximum values of your dataset
- mean: the mean of your dataset
- variance: the variance of your dataset
- skewness: the skewness of your dataset
- kurtosis: the kurtosis of your dataset

You can access particular values with dot notation:

In [74]:
result.nobs

result.minmax[0]  # Min

result.minmax[1]  # Max

result.mean

result.variance

result.skewness

result.kurtosis


9

-5.0

41.0

11.622222222222222

228.75194444444446

0.9249043136685094

0.14770623629658886

In [76]:
#pandas

result = z.describe()
result


result['mean']

result['std']

result['min']

result['max']

result['25%']

result['50%']

result['75%']


count     9.000000
mean     11.622222
std      15.124548
min      -5.000000
25%       0.100000
50%       8.000000
75%      21.000000
max      41.000000
dtype: float64

11.622222222222222

15.12454774346805

-5.0

41.0

0.1

8.0

21.0

### Measures of Correlation Between Pairs of Data

You’ll often need to examine the relationship between the corresponding elements of two variables in a dataset. Say there are two variables, 𝑥 and 𝑦, with an equal number of elements, 𝑛. Let 𝑥₁ from 𝑥 correspond to 𝑦₁ from 𝑦, 𝑥₂ from 𝑥 to 𝑦₂ from 𝑦, and so on. You can then say that there are 𝑛 pairs of corresponding elements: (𝑥₁, 𝑦₁), (𝑥₂, 𝑦₂), and so on.

You’ll see the following measures of correlation between pairs of data:

Positive correlation exists when larger values of 𝑥 correspond to larger values of 𝑦 and vice versa.
Negative correlation exists when larger values of 𝑥 correspond to smaller values of 𝑦 and vice versa.
Weak or no correlation exists if there is no such apparent relationship.



The two statistics that measure the correlation between datasets are covariance and the correlation coefficient. Let’s define some data to work with these measures. You’ll create two Python lists and use them to get corresponding NumPy arrays and Pandas Series:

In [77]:
x = list(range(-10, 11))

y = [0, 2, 2, 2, 2, 3, 3, 6, 7, 4, 7, 6, 6, 9, 4, 5, 5, 10, 11, 12, 14]

x_, y_ = np.array(x), np.array(y)

x__, y__ = pd.Series(x_), pd.Series(y_)

### Covariance
The sample covariance is a measure that quantifies the strength and direction of a relationship between a pair of variables:

If the correlation is positive, then the covariance is positive, as well. A stronger relationship corresponds to a higher value of the covariance.
If the correlation is negative, then the covariance is negative, as well. A stronger relationship corresponds to a lower (or higher absolute) value of the covariance.
If the correlation is weak, then the covariance is close to zero.
The covariance of the variables 𝑥 and 𝑦 is mathematically defined as 𝑠ˣʸ = Σᵢ (𝑥ᵢ − mean(𝑥)) (𝑦ᵢ − mean(𝑦)) / (𝑛 − 1), where 𝑖 = 1, 2, …, 𝑛, mean(𝑥) is the sample mean of 𝑥, and mean(𝑦) is the sample mean of 𝑦. It follows that the covariance of two identical variables is actually the variance: 𝑠ˣˣ = Σᵢ(𝑥ᵢ − mean(𝑥))² / (𝑛 − 1) = (𝑠ˣ)² and 𝑠ʸʸ = Σᵢ(𝑦ᵢ − mean(𝑦))² / (𝑛 − 1) = (𝑠ʸ)².

This is how you can calculate the covariance in pure Python:

In [78]:
n = len(x)

mean_x, mean_y = sum(x) / n, sum(y) / n

cov_xy = (sum((x[k] - mean_x) * (y[k] - mean_y) for k in range(n))
         / (n - 1))
cov_xy

19.95

In [79]:
#NumPy has the function cov() that returns the covariance matrix:

cov_matrix = np.cov(x_, y_)
cov_matrix

array([[38.5       , 19.95      ],
       [19.95      , 13.91428571]])

In [80]:
# Note that cov() has the optional parameters bias, which defaults to False, and ddof, which defaults to None. 
# Their default values are suitable for getting the sample covariance matrix. 
# The upper-left element of the covariance matrix is the covariance of x and x, or the variance of x. 
# Similarly, the lower-right element is the covariance of y and y, or the variance of y. 
# You can check to see that this is true:

x_.var(ddof=1)

y_.var(ddof=1)

38.5

13.914285714285711

As you can see, the variances of x and y are equal to cov_matrix[0, 0] and cov_matrix[1, 1], respectively.

The other two elements of the covariance matrix are equal and represent the actual covariance between x and y:

In [81]:
cov_xy = cov_matrix[0, 1]
cov_xy

cov_xy = cov_matrix[1, 0]
cov_xy

19.95

19.95

You’ve obtained the same value of the covariance with np.cov() as with pure Python.

Pandas Series have the method .cov() that you can use to calculate the covariance:

### Correlation Coefficient

The correlation coefficient, or Pearson product-moment correlation coefficient, is denoted by the symbol 𝑟. The coefficient is another measure of the correlation between data. You can think of it as a standardized covariance. Here are some important facts about it:

The value 𝑟 > 0 indicates positive correlation.
The value 𝑟 < 0 indicates negative correlation.

The value r = 1 is the maximum possible value of 𝑟. It corresponds to a perfect positive linear relationship between variables.

The value r = −1 is the minimum possible value of 𝑟. It corresponds to a perfect negative linear relationship between variables.

The value r ≈ 0, or when 𝑟 is around zero, means that the correlation between variables is weak.
The mathematical formula for the correlation coefficient is 𝑟 = 𝑠ˣʸ / (𝑠ˣ𝑠ʸ) where 𝑠ˣ and 𝑠ʸ are the standard deviations of 𝑥 and 𝑦 respectively. If you have the means (mean_x and mean_y) and standard deviations (std_x, std_y) for the datasets x and y, as well as their covariance cov_xy, then you can calculate the correlation coefficient with pure Python:

In [82]:
var_x = sum((item - mean_x)**2 for item in x) / (n - 1)

var_y = sum((item - mean_y)**2 for item in y) / (n - 1)

std_x, std_y = var_x ** 0.5, var_y ** 0.5

r = cov_xy / (std_x * std_y)

r

0.861950005631606

In [83]:
# You’ve got the variable r that represents the correlation coefficient.

# scipy.stats has the routine pearsonr() that calculates the correlation coefficient and the 𝑝-value:

r, p = scipy.stats.pearsonr(x_, y_)

r
p

0.861950005631606

5.122760847201171e-07

pearsonr() returns a tuple with two numbers. The first one is 𝑟 and the second is the 𝑝-value.

Similar to the case of the covariance matrix, you can apply np.corrcoef() with x_ and y_ as the arguments and get the correlation coefficient matrix:

In [84]:
corr_matrix = np.corrcoef(x_, y_)

corr_matrix

array([[1.        , 0.86195001],
       [0.86195001, 1.        ]])

The upper-left element is the correlation coefficient between x_ and x_. The lower-right element is the correlation coefficient between y_ and y_. Their values are equal to 1.0. The other two elements are equal and represent the actual correlation coefficient between x_ and y_:

In [85]:
r = corr_matrix[0, 1]
r

r = corr_matrix[1, 0]
r

0.8619500056316061

0.861950005631606

In [86]:
#Of course, the result is the same as with pure Python and pearsonr().

#You can get the correlation coefficient with scipy.stats.linregress():

scipy.stats.linregress(x_, y_)

LinregressResult(slope=0.5181818181818181, intercept=5.714285714285714, rvalue=0.861950005631606, pvalue=5.122760847201164e-07, stderr=0.06992387660074979, intercept_stderr=0.4234100995002589)

linregress() takes x_ and y_, performs linear regression, and returns the results. slope and intercept define the equation of the regression line, while rvalue is the correlation coefficient. To access particular values from the result of linregress(), including the correlation coefficient, use dot notation:

In [87]:
result = scipy.stats.linregress(x_, y_)
r = result.rvalue
r

0.861950005631606

In [88]:
#That’s how you can perform linear regression and obtain the correlation coefficient.

#Pandas Series have the method .corr() for calculating the correlation coefficient:

r = x__.corr(y__)
r

r = y__.corr(x__)
r

0.8619500056316061

0.861950005631606